### PR TITLE
Support SA Events 

### PR DESCRIPTION
**Pull request by Jivings**
Added support for Events with a few extra config params;

gasyby-config.js
```
plugins: [    
    {
      resolve: 'simple-analytics-gatsby-plugin',
      options: {
        eventsGlobal: 'sa',
        events: true,
        trackPageViews: true
      }
    },
```

- `eventsGlobal` overrides the default `window.sa` global.
- Setting `trackPageViews` to true tracks Gatsby Route Updates as events like this;

gatsby-browser.js
```
exports.onRouteUpdate = ({ location }, pluginOptions) => {
  if (pluginOptions.trackPageViews) {
    const ignore = pluginOptions.ignorePages || []
    const { pathname } = location
    if (!ignore.includes(pathname)) {
      sa(pathname)
    }
  }
}
``` 

- I exposed the tracking function as a es import so instead of using `window.sa` you can import the plugin script and use that;

```
import sa from 'gatsby-plugin-simple-analytics';

sa('signed-up');
```

- For convenience I added a small debug function to the index that prints debug logs when running in dev mode.

I can't test all of this right now because I'm stuck with the issues I sent to you in Telegram so it's not ready to merge.

### CODE REVIEW

**Comment by Jivings** [comment]
Sorry once you've merged the other PR then this wont have all that PRs commits in it too 😅

File: src/gatsby-ssr.js [hunk 1]
```diff
@@ -1,23 +1,71 @@
...
-    defer: true
+    async: true
   }
 
   if (pluginOptions.metomic) {
     options.type = 'text/x-metomic'
-    options['data-micropolicy'] = pluginOptions.metomic
+    options.metomic = pluginOptions.metomic
+  }
+  if (pluginOptions.events) {
```
Comment (adriaanvanrossum) [inline src/gatsby-ssr.js]: Should we make this `trackEvents` to be consistent with `trackPageViews`?

File: src/gatsby-ssr.js [hunk 2]
```diff
@@ -1,23 +1,71 @@
...
+    options.eventsGlobal = pluginOptions.eventsGlobal
   }
 
   setHeadComponents([
-    React.createElement('script', options)
+    React.createElement('script', {
+      key: 'plugin-simpleanalytics',
+      dangerouslySetInnerHTML: {
```
Comment (adriaanvanrossum) [inline src/gatsby-ssr.js]: This always looks scary ;-)

File: src/gatsby-ssr.js [hunk 3]
```diff
@@ -1,23 +1,71 @@
...
 }
+
+// this script runs when the page loads, it does the following;
+// 1. Checks the doNotTrack option, and if set then it aborts and does nothing
+// 2. Creates a script tag for the Simple Analytics script
+// 3. Sets all the options properties on the script tag
+// 4. Inserts the script tag before the first other script in the document
+const loadScript = (domain, options) => {
+  return `!(function(s, i, m, p, l, e) {
+    const dnt = s.doNotTrack || m.doNotTrack || m.msDoNotTrack;
```
Comment (adriaanvanrossum) [inline src/gatsby-ssr.js]: `const` is not supported in every browser, or will this be compiled as well?

File: src/index.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+const onLoad = function() {
+  let event
+  while ((event = eventBuffer.pop())) {
+    trackEvent(event)
+  }
+}
+
+const trackEvent = function(event) {
+  if (!eventGlobal) {
+    const script = document.querySelector('#simple-analytics')
```
Comment (adriaanvanrossum) [inline src/index.js]: To make it a bit more safe (maybe convert to ES5 if not comnpiled):

```html
const script = document.querySelector(`[src="${options.src}"]`)
```

**Comment by adriaanvanrossum** [comment]
A though I have: should we include the tracking script itself in the application. I want to add versioning to our scripts so people can use SRI (https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity). If we use versioning it makes sense to include the tracking scripts in the app code as well. Why not? What is your opinion on this?

File: src/gatsby-ssr.js (hunk 3 above)
Comment (Jivings) [reply src/gatsby-ssr.js]: Yeah it's compiled with babel.

**Comment by Jivings** [comment]
I have a bit of a problem when using this with the Metomic code.

Because Metomic changes the script type to `"text/x-metomic"` instead of `"text/javascript"`, the onload event is never fired so we lose some events if they take place before the script is loaded. 

Just trying to figure out a way around that...

**Comment by adriaanvanrossum** [comment]
Cool, let me know if I can be of any help.

**Comment by adriaanvanrossum** [comment]
Is this ready to be merged?

**Comment by Jivings** [comment]
Not quite, I've been testing it out on leavemealone.app [1] for the last couple of days and there are a couple of things I want to check. Getting some events though!

[1]: https://simpleanalytics.com/events/leavemealone.app

**Comment by tmm** [comment]
@Jivings great work here! need some help getting it across the finish line? i would love to use.

**Comment by adriaanvanrossum** [comment]
Let me know if you need any help from me!

**Comment by Jivings** [comment]
@tmm was waiting for the SA events code to be part of the main bundle so that I could simplify this a little. 

Are they now combined @adriaanvanrossum?

**Comment by adriaanvanrossum** [comment]
Yes, they are now: https://scripts.simpleanalyticscdn.com/latest.js

**Comment by Jivings** [comment]
Okay will update!

**Comment by adriaanvanrossum** [comment]
Cool! Also used your code a bit for this one: https://github.com/simpleanalytics/vuepress-plugin/blob/master/enhanceAppFile.js

**Comment by Jivings** [comment]
Pushed, but I will test on subscriptionscore.com for a few hours and make sure is all good 👍

File: .prettierrc [hunk 5]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (adriaanvanrossum) [inline .prettierrc]: We are using prettier defaults, for everything. It's pretty great!

File: src/gatsby-ssr.js [hunk 6]
```diff
@@ -1,16 +1,14 @@
...
+const defaultDomain = 'scripts.simpleanalyticscdn.com'
 
 exports.onRenderBody = ({ setHeadComponents }, pluginOptions) => {
   const domain = pluginOptions.domain || defaultDomain
   let scriptName
   if (pluginOptions.scriptName) {
     scriptName = pluginOptions.scriptName
-  } else if (pluginOptions.trackEvents) {
-    scriptName = 'e.js'
   } else if (domain === defaultDomain) {
```
Comment (adriaanvanrossum) [inline src/gatsby-ssr.js]: It's now `latest.js` for both! No need to check anymore.

**Comment by adriaanvanrossum** [comment]
> will test on subscriptionscore.com for a few hours

Cool, let me know how that goes. Don't forget to use `latest.js` if you want to use events.

**Comment by Jivings** [comment]
Okay seems to work using `latest.js` and my custom domain.

I'm happy for it to get merged. 

image

**Comment by adriaanvanrossum** [comment]
Latest version is now live: https://www.npmjs.com/package/gatsby-plugin-simple-analytics

Thanks @Jivings! ❤️

**Pull request merged**